decompiler/structures/pseudo/complextypes.py

```diff
@@ -0,0 +1,139 @@
+        self.members[member.offset] = member
+
+    def get_member_by_offset(self, offset: int) -> ComplexTypeMember:
+        return self.members.get(offset)
+
+    def declaration(self) -> str:
+        members = ";\n\t".join(self.members[k].declaration() for k in sorted(self.members.keys())) + ";"
+        return f"{self.type_specifier.value} {self.name} {{\n\t{members}\n}}"
+
+
+@dataclass(frozen=True, order=True)
+class Union(ComplexType):
+    members: List[ComplexTypeMember] = field(compare=False)
+    type_specifier = ComplexTypeSpecifier.UNION
+
+    def add_member(self, member: ComplexTypeMember):
+        self.members.append(member)
+
+    def declaration(self) -> str:
+        members = ";\n\t".join(x.declaration() for x in self.members) + ";"
+        return f"{self.type_specifier.value} {self.name} {{\n\t{members}\n}}"
+
+    def get_member_by_type(self, _type: Type) -> ComplexTypeMember:
+        """Retrieve member of union by its type."""
+        for member in self.members:
+            if member.type == _type:
+                return member
+
+
+@dataclass(frozen=True, order=True)
+class Enum(ComplexType):
+    members: Dict[int, ComplexTypeMember] = field(compare=False)
+    type_specifier = ComplexTypeSpecifier.ENUM
+
+    def add_member(self, member: ComplexTypeMember):
+        self.members[member.value] = member
+
+    def get_name_by_value(self, value: int) -> str:
+        return self.members.get(value).name
+
+    def declaration(self) -> str:
+        members = ",\n\t".join(f"{x.name} = {x.value}" for x in self.members.values())
+        return f"{self.type_specifier.value} {self.name} {{\n\t{members}\n}}"
+
+
+@dataclass(frozen=True, order=True)
+class ComplexTypeName(Type):
+    """Class that store a name of a complex type. Used to prevent recursions when constructing
+    struct(...) members of the same complex type"""
+
+    name: str
+
+    def __str__(self) -> str:
+        return self.name
+
+
+class ComplexTypeMap:
+    """A class in charge of storing complex custom/user defined types by their string representation"""
+
+    def __init__(self):
+        self._name_to_type_map: Dict[ComplexTypeName, ComplexType] = {}
+
+    def retrieve_by_name(self, typename: ComplexTypeName) -> ComplexType:
+        """Get complex type by name; used to avoid recursion."""
+        return self._name_to_type_map.get(typename, None)
+
+    def add(self, complex_type: ComplexType):
+        """Add complex type to the mapping."""
+        self._name_to_type_map[ComplexTypeName(0, complex_type.name)] = complex_type
+
+    def pretty_print(self):
+        for t in self._name_to_type_map.values():
+            logging.error(t.declaration())
+
+    def declarations(self) -> str:
+        """Returns declarations of all complex types used in decompiled function."""
+        return ";\n".join(t.declaration() for t in self._name_to_type_map.values()) + ";" if self._name_to_type_map else ""
```

Comment:
Kinda seems like a TypedDict - or did I miss something here?

Reply:
I was just wondering if there is any functionality here aside from this class being `Dict[ComplexTypeName, ComplexType]` but I guess we can keep it to have it future proof